### PR TITLE
DXE-6751 Add improved build script for on-demand-ci

### PR DESCRIPTION
**Pull request by Wajktor13**

### CODE REVIEW

File: .github/workflows/on-demand-ci.yml [hunk 1]
```diff
@@ -6,9 +6,10 @@ on:
     inputs:
       variants:
         description: |
-          Comma-separated variants to build.
+          Space-separated image names to build (dependency chains are resolved automatically).
           Leave empty to build all variants from the variants file.
-          Example: "base,base cli,cli appsec,appsec shell" builds shell with appsec.
+          Example: "purge dns" builds purge and dns.
```
Comment (artbookspirit) [inline .github/workflows/on-demand-ci.yml]: If you run the python script for purge and dns, ("base", "cli") will be built twice.

I understand that this is not a problem - that a cached version will be picked up.

(Just asking to be sure).

File: scripts/build.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    if ":" in spec:
+        _, targets = spec.split(":", 1)
+        deps = [s.strip() for s in targets.split(",") if s.strip()]
+
+    for d in deps:
+        if d not in VARIANTS:
+            sys.exit(f"error: unknown dependency '{d}'")
+
+    # Flatten and deduplicate preserving order via dict keys
+    calls = list(dict.fromkeys(c for d in deps for c in VARIANTS[d]))
```
Comment (artbookspirit) [inline scripts/build.py]: This line is a bit too clever. I'd expand it to a regular nested for with deduplication via set.

File: scripts/build.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    else:
+        sys.exit(f"error: unknown image '{spec}'")
+
+    print(f"Building '{spec.split(':', 1)[0]}': {' -> '.join(c[-1] for c in calls)}")
+    for c in calls:
+        run_build_chain(c, args.dry_run)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (artbookspirit) [inline scripts/build.py:105]: Optional: Please consider writing tests using pytest. 

Tests can easily parse output when dry run is enabled - no need for anything more fancy imo.

**Review by artbookspirit — CHANGES_REQUESTED** [review]

File: .github/workflows/on-demand-ci.yml (hunk 1 above)
Comment (Wajktor13) [reply .github/workflows/on-demand-ci.yml]: I think it's not an issue - as you said cache should work here.

File: scripts/build.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        epilog="examples:\n  %(prog)s base\n  %(prog)s shell\n  %(prog)s \"shell:terraform,api-gateway\""
+    )
+    parser.add_argument("image", metavar="IMAGE", help="Image to build. Supports 'shell:img1,img2'.")
+    parser.add_argument("--dry-run", action="store_true", help="Print commands without executing.")
+    args = parser.parse_args()
+
+    spec = args.image
+    if spec == "shell" or spec.startswith("shell:"):
+        calls = resolve_shell_chain(spec)
+    elif spec in VARIANTS:
```
Comment (wzagrajcz) [inline scripts/build.py:94]: Does it work when mutliple images are provided? In the example in yaml, there is `Example: "purge dns" builds purge and dns.`. Shouldn't we have here a list from the specs separated by space and get the list of calls from list of variants similarly as there is in shell? If we don't want to support multiple images (considering that it can be worked around by having that simply added into shell in case) then the example needs correction.

File: scripts/build.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+"""Build akamai-docker images with automatic dependency resolution."""
+
+import argparse
+import subprocess
+import sys
+
+VARIANTS: dict[str, list[tuple[str, ...]]] = {
```
Comment (wzagrajcz) [inline scripts/build.py:8]: This solution looks very good considering that it gives both simplicty and elasticity. Problem is that long term we will have two sources of truth for defining dependencies between images (here + variants). That's probably topic for different story (since this one is already way bigger then expected), but imo it'd be good to have it unified. WDYT?

**Review by wzagrajcz — CHANGES_REQUESTED** [review]

File: scripts/build.py (hunk 5 above)
Comment (Wajktor13) [reply scripts/build.py:8]: It should definitely be unified

File: scripts/build.py (hunk 3 above)
Comment (Wajktor13) [reply scripts/build.py:105]: added

File: scripts/build.py (hunk 4 above)
Comment (artbookspirit) [reply scripts/build.py:94]: For me personally, this is enough to build either:

- a single image except shell
- shell with default dependencies or selected dependencies.

wdyt?

Comment (Wajktor13) [reply scripts/build.py:94]: It does work, but in such case the script is run twice - once with `purge` argument, and once with `dns`.

Example of such run: https://github.com/akamai/akamai-docker/actions/runs/27823363305/job/82341569302

File: scripts/test_build.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    def test_conflict_message_includes_both_paths(self):
+        calls = [("a", "target"), ("b", "target")]
+        with pytest.raises(SystemExit) as exc_info:
+            validate_no_conflicting_paths(calls)
+        msg = str(exc_info.value)
+        assert "('a', 'target')" in msg
+        assert "('b', 'target')" in msg
+
+
+class TestResolveShellChain:
```
Comment (artbookspirit) [inline scripts/test_build.py]: Well, I'm not a fan of such low-level tests, to be honest. TestValidateNoConflictingPaths is OK, but the following tests contain a lot of creativity which makes them hard to read while not testing the most basic thing which is the exact output of the tool.

Why not just crate a single table test that takes the input, executes the tool with dry run, parses the output and compares the output with expected output which is hard coded into the test?

For fancy corner-cases, we can write an additional small test with 3-5 cases max.

**Review by artbookspirit — CHANGES_REQUESTED** [review]

File: scripts/build.py (hunk 4 above)
Comment (wzagrajcz) [reply scripts/build.py:94]: I've talked to Wiktor about this. I didn't noticed that the multiple images is handleded in the `on-demand-ci.yml` file. So currently it should work fine and I do not see any complexity problems here tbh. So imo it can stay as it is, but if you see any reasons why it should be limited to what Piotr's proposed above, I'm also fine with it.

File: scripts/build.py (hunk 5 above)
Comment (Wajktor13) [reply scripts/build.py:8]: Story created - DXE-6831

**Review by wzagrajcz — APPROVED** [review]

File: scripts/test_build.py (hunk 6 above)
Comment (Wajktor13) [reply scripts/test_build.py]: Changed

File: scripts/build.py (hunk 4 above)
Comment (Wajktor13) [reply scripts/build.py:94]: Applied the Piotr's suggestion

**Review by artbookspirit — APPROVED** [review]
